test/system/065-cp.bats

```diff
@@ -7,6 +7,312 @@
+    run_podman exec cpcontainer cat /tmp/hostfile2
+    is "$output" "This file is on the host (2)"
+
+    run_podman cp $srcdir/hostfile cpcontainer:/tmp/anotherbase.txt
+    run_podman exec cpcontainer cat /tmp/anotherbase.txt
+    is "$output" "This file is on the host"
+
+    # Copy to a relative path -> must resolve to the container's work dir
+    # (i.e., /srv).
+    run_podman exec cpcontainer mkdir /srv/subdir
+    run_podman cp $srcdir/hostfile cpcontainer:subdir
+    run_podman exec cpcontainer cat /srv/subdir/hostfile
+    is "$output" "This file is on the host"
+
+    run_podman cp $srcdir/hostfile cpcontainer:anotherbase.txt
+    run_podman exec cpcontainer cat /srv/anotherbase.txt
+    is "$output" "This file is on the host"
+
+    # Host path does not exist.
+    run_podman 125 cp $srcdir/IdoNotExist cpcontainer:/tmp
+    is "$output" 'Error: ".*/IdoNotExist" could not be found on the host'
+
+    # Container path does not exist.  Notice that the error message shows how
+    # the specified container is resolved.
+    run_podman 125 cp $srcdir/hostfile cpcontainer:/IdoNotExist/
+    is "$output" 'Error: "/IdoNotExist/" could not be found on container.*(resolved to .*/IdoNotExist.*'
+
+    run_podman rm -f cpcontainer
+}
+
+
+@test "podman cp --extract=true tar archive to container" {
+    skip_if_remote "podman-remote does not yet handle cp"
+
+    # Create tempfile with random name and content
+    srcdir=$PODMAN_TMPDIR/cp-test-stdin
+    mkdir -p $srcdir
+    rand_filename=$(random_string 20)
+    rand_content=$(random_string 50)
+    echo $rand_content > $srcdir/$rand_filename
+    chmod 644 $srcdir/$rand_filename
+
+    # Now tar it up!
+    tar_file=$PODMAN_TMPDIR/archive.tar.gz
+    tar -zvcf $tar_file $srcdir
+
+    run_podman run -d --name cpcontainer $IMAGE sleep infinity
+
+    # First just copy without extracting the archive.
+    run_podman cp $tar_file cpcontainer:/tmp
+    # Now remove the archive which will also test if it exists and is a file.
+    # To save expensive exec'ing, create a file for the next tests.
+    run_podman exec cpcontainer sh -c "rm /tmp/archive.tar.gz; touch /tmp/file.txt"
+
+    # Now copy with extracting the archive. NOTE that Podman should
+    # auto-decompress the file if needed.
+    run_podman cp --extract=true $tar_file cpcontainer:/tmp
+    run_podman exec cpcontainer cat /tmp/$srcdir/$rand_filename
+    is "$output" "$rand_content"
+
+    # Test extract on non archive.
+    run_podman cp --extract=true $srcdir/$rand_filename cpcontainer:/foo.txt
+
+    # Cannot extract an archive to a file!
+    run_podman 125 cp --extract=true $tar_file cpcontainer:/tmp/file.txt
+    is "$output" 'Error: cannot extract archive .* to file "/tmp/file.txt"'
+
+    run_podman rm -f cpcontainer
+}
+
+
+@test "podman cp file from container to host" {
+    skip_if_remote "podman-remote does not yet handle cp"
+
+    srcdir=$PODMAN_TMPDIR/cp-test-file-ctr-to-host
+    mkdir -p $srcdir
+
```

Comment:
I meant to make time to refactor this, too, into table-driven form; but time got away from me. Given the example above, would you mind using that as a template to refactor this and the next two?

Reply:
Sure, thanks!